### PR TITLE
 General Storage + Login

### PR DESCRIPTION
**Pull request by FrederikBolding**

### CODE REVIEW

**Comment by FrederikBolding** [comment]
[ch6995]

File: src/app/screens/NewUser.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    if (!result) {
+      setError('An error occurred');
+    }
+  };
+
+  return (
+    <div>
+      Welcome!
+      <br />
+      <label htmlFor="password">
```
Comment (Mrtenz) [inline src/app/screens/NewUser.tsx:25]: You don't need a `htmlFor` property if the input is inside the label.

File: src/app/screens/NewUser.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        Enter a new master password
+        <input
+          id="password"
+          name="password"
+          type="password"
+          onChange={(e) => setPassword(e.currentTarget.value)}
```
Comment (Mrtenz) [inline src/app/screens/NewUser.tsx:31]: Don't use arrow functions inside properties, they cause unnecessary re-renders.

https://stackoverflow.com/a/36677798/11285318

File: src/app/screens/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  const [password, setPassword] = useState('');
+  const [error, setError] = useState('');
+  const dispatch = useDispatch();
+
+  const handleLogin = async () => {
+    const result = await login(password);
+    dispatch(setLoginState(result ? LoginState.LOGGED_IN : LoginState.LOGGED_OUT));
+    if (!result) {
+      setError('An error occurred');
+    }
```
Comment (Mrtenz) [inline src/app/screens/Login.tsx:18]: Maybe handle Promise rejections here as well.

File: src/app/screens/NewUser.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  const [password, setPassword] = useState('');
+  const [error, setError] = useState('');
+  const dispatch = useDispatch();
+
+  const handleCreate = async () => {
+    const result = await init(password);
+    dispatch(setLoginState(result ? LoginState.LOGGED_IN : LoginState.NEW_USER));
+    if (!result) {
+      setError('An error occurred');
+    }
```
Comment (Mrtenz) [inline src/app/screens/NewUser.tsx:18]: Maybe handle Promise rejections here as well.

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: src/api/db.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    case DBRequestType.GET_LOGIN_STATE:
+      return getLoginState();
+    case DBRequestType.GET_ACCOUNTS:
+      return Promise.resolve(getAccounts());
+    default:
+      throw new Error('Undefined request type');
+  }
+};
+
+export const testables = { handleRequest };
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/api/db.ts:69]: I just wanted to export it for testing, but didn't want it easily accessible from the outside.

File: src/types/db.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+export enum DBRequestType {
+  INIT,
+  LOGIN,
+  GET_LOGIN_STATE,
+  GET_ACCOUNTS
+}
+
+export enum LoginState {
+  NEW_USER,
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/types/db.ts:13]: I get that. We can use two booleans instead, but if the user has never set up the app, telling whether they're `loggedIn` is meaningless. That's why I felt that using an enum to represent the 3 states was better.

File: src/api/db.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import { app, ipcMain } from 'electron';
+import Store from 'electron-store';
+import fs from 'fs';
+import path from 'path';
+
+import { IPC_CHANNELS } from '@config';
+import { DBRequest, DBRequestType, DBResponse, IAccount, LoginState } from '@types';
+
+let store: Store;
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply src/api/db.ts:9]: Instead of using `electron-store`?

[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/api/db.ts:9]: We can still use `electron-store` (for persisting), but I think a Redux store in between is more scalable.

File: src/api/db.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/api/db.ts:69]: I would personally just export the function. It's a private API regardless, so there's no significant difference.

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**